krm-functions/dnn-fn/main.go

```diff
@@ -0,0 +1,194 @@
+			Owns: map[corev1.ObjectReference]kptcondsdk.ResourceKind{
+				{
+					APIVersion: ipamv1alpha1.GroupVersion.Identifier(),
+					Kind:       ipamv1alpha1.IPAllocationKind,
+				}: kptcondsdk.ChildRemote,
+			},
+			Watch: map[corev1.ObjectReference]kptcondsdk.WatchCallbackFn{
+				{
+					APIVersion: infrav1alpha1.GroupVersion.Identifier(),
+					Kind:       GetKindOrPanic[infrav1alpha1.ClusterContext](),
+				}: myFn.ClusterContextCallbackFn,
+			},
+			PopulateOwnResourcesFn: myFn.desiredOwnedResourceList,
+			GenerateResourceFn:     myFn.updateDnnResource,
+		},
+	)
+	if err != nil {
+		rl.Results.ErrorE(err)
+		return false, nil
+	}
+	return myFn.sdk.Run()
+}
+
+func (f *DnnFn) ClusterContextCallbackFn(o *fn.KubeObject) error {
+	var cluster infrav1alpha1.ClusterContext
+	err := o.As(&cluster)
+	if err != nil {
+		return err
+	}
+	if cluster.Spec.SiteCode == nil || *cluster.Spec.SiteCode == "" {
+		return fmt.Errorf("mandatory field `siteCode` is missing from ClusterContext %q", cluster.Name)
+	}
+	if f.site != "" && f.site != *cluster.Spec.SiteCode {
+		return fmt.Errorf("multiple ClusterContext objects with confliciting `siteCode` fields found in the package")
+	}
+	f.site = *cluster.Spec.SiteCode
+	return nil
+}
+
+func (f *DnnFn) desiredOwnedResourceList(o *fn.KubeObject) (fn.KubeObjects, error) {
+	resources := fn.KubeObjects{}
+
+	// get "parent"| DNN struct
+	var dnn nephioreqv1alpha1.DataNetwork
+	err := o.As(&dnn)
+	if err != nil {
+		return nil, err
+	}
+
+	// add IPAllocation for each pool
+	for _, pool := range dnn.Spec.Pools {
+		ipalloc := ipamv1alpha1.BuildIPAllocation(
+			metav1.ObjectMeta{
+				Name: fmt.Sprintf("%s-%s", dnn.Name, pool.Name),
+			},
+			ipamv1alpha1.IPAllocationSpec{
+				PrefixKind: ipamv1alpha1.PrefixKindPool,
+				NetworkInstance: &corev1.ObjectReference{
+					Name: dnn.Spec.NetworkInstance.Name,
+				},
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						ipam_common.NephioSiteKey: f.site,
+					},
+				},
+				PrefixLength: pool.PrefixLength,
+			},
+			ipamv1alpha1.IPAllocationStatus{},
+		)
+
+		ipallocObj, err := fn.NewFromTypedObject(ipalloc)
+		if err != nil {
+			return nil, err
+		}
+		resources = append(resources, ipallocObj)
+	}
+	return resources, nil
+}
+
+func (f *DnnFn) updateDnnResource(dnnObj *fn.KubeObject, owned fn.KubeObjects) (*fn.KubeObject, error) {
+	if dnnObj == nil {
+		return nil, fmt.Errorf("expected a DNN object, but got nil")
+	}
+	var pools []DnnPoolStatus
+	for _, o := range owned.Where(fn.IsGroupVersionKind(ipamv1alpha1.IPAllocationGroupVersionKind)) {
+		var ipalloc ipamv1alpha1.IPAllocation
+		err := o.As(&ipalloc)
+		if err != nil {
+			return nil, err
+		}
+
+		poolName, found := strings.CutPrefix(ipalloc.Name, dnnObj.GetName()+"-")
+		if found {
+			pools = append(pools, DnnPoolStatus{Name: poolName, Ip: ipalloc.Status})
+		} else {
+			f.rl.Results.Warningf("found an owned IPAllocation with a suspicious name: %v", ipalloc.Name)
```

Comment:
is it not better to log rather than adding to resourceList? I would like to avoid adding to resourceList.